### PR TITLE
[WIP] Add a GLFW backend for NoGUI

### PR DESCRIPTION
**Pull request by linkmauve**
This PR is very incomplete and only here for CI, but can already use the Vulkan renderer on both X11 and Wayland (and I guess on OSX and Windows too).

Input is still missing altogether, as is GL rendering, and proper GL context creation.

### CODE REVIEW

File: Source/Core/DolphinNoGUI/GLFW.h [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+
+#include "Core/ConfigManager.h"
+#include "Core/Core.h"
+#include "Core/State.h"
+
+#include "DolphinNoGUI/Platform.h"
+
+static void OnError(int error, const char* description)
+{
+  fprintf(stderr, "GLFW: 0x%08x: %s\n", error, description);
```
Comment (BhaaLseN) [inline Source/Core/DolphinNoGUI/GLFW.h]: I'm pretty sure `fprintf` needs `stdio.h`/ `cstdio``  to be included.

File: Source/Core/DolphinNoGUI/GLFW.h [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+  else if (key >= GLFW_KEY_F1 && key <= GLFW_KEY_F8)
+  {
+    int slot_number = key - GLFW_KEY_F1 + 1;
+    if (mods & GLFW_MOD_SHIFT)
+      State::Save(slot_number);
+    else
+      State::Load(slot_number);
+  }
+  else if (key == GLFW_KEY_F9)
+    Core::SaveScreenShot();
```
Comment (BhaaLseN) [inline Source/Core/DolphinNoGUI/GLFW.h]: There are blocks with braces, so those should have braces too (even if they're just one-liners)

File: Source/Core/DolphinNoGUI/GLFW.h [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+  {
+    glfwSetErrorCallback(OnError);
+
+    if (!glfwInit())
+    {
+      PanicAlert("GLFW initialization failed");
+      exit(1);
+    }
+
+#define LET_S_USE_OPENGL
```
Comment (BhaaLseN) [inline Source/Core/DolphinNoGUI/GLFW.h]: This looks a little like testing code, with the `#define` right next to its usage. Should this be towards the top of the file, or in the CMake files instead?

File: Source/Core/VideoBackends/Vulkan/SwapChain.cpp [hunk 4]
```diff
@@ -39,7 +41,20 @@ SwapChain::~SwapChain()
...
 {
-#if defined(VK_USE_PLATFORM_WIN32_KHR)
+#if defined(HAVE_GLFW)
+  GLFWwindow* window = static_cast<GLFWwindow*>(hwnd);
+  VkSurfaceKHR surface;
+
+  VkResult res = glfwCreateWindowSurface(instance, window, nullptr, &surface);
+  if (res != VK_SUCCESS)
+  {
+    LOG_VULKAN_ERROR(res, "glfwCreateWindowSurface failed: ");
```
Comment (BhaaLseN) [inline Source/Core/VideoBackends/Vulkan/SwapChain.cpp]: Is this just a suffix, or is there some output missing? Looks like half an error message to me, but I don't know the behavior behind those functions.

File: Source/Core/DolphinNoGUI/Platform.h [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  virtual void MainLoop()
+  {
+    while (s_running.IsSet())
+    {
+      Core::HostDispatchJobs();
+      std::this_thread::sleep_for(std::chrono::milliseconds(100));
+    }
+  }
+  virtual void Shutdown() {}
+  virtual ~Platform() {}
```
Comment (lioncash) [inline Source/Core/DolphinNoGUI/Platform.h]: Constructor/Destructor should be the first thing in public interfaces.

File: Source/Core/DolphinNoGUI/Platform.h [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+#pragma once
+
+#include <string>
+
+class Platform
+{
+public:
+  virtual void Init() {}
+  virtual void SetTitle(const std::string& title) {}
+  virtual void MainLoop()
```
Comment (lioncash) [inline Source/Core/DolphinNoGUI/Platform.h]: This should go in the cpp file. If this ever changes, everything that includes this file will need to be rebuilt as is.

`Core/Core.h`, `<chrono>` and `<thread>` would also need to be included.

File: Source/Core/DolphinNoGUI/GLFW.h [hunk 7]
```diff
@@ -0,0 +1,211 @@
+// Copyright 2016 Dolphin Emulator Project
```
Comment (lioncash) [inline Source/Core/DolphinNoGUI/GLFW.h]: This file should probably have a cpp file that all the implementation details go into.

File: Source/Core/VideoBackends/Vulkan/VulkanContext.cpp [hunk 8]
```diff
@@ -172,6 +176,21 @@ bool VulkanContext::SelectInstanceExtensions(ExtensionList* extension_list, bool
     return true;
   };
 
+#if HAVE_GLFW
+
+  uint32_t required_extension_count = 0;
```
Comment (lioncash) [inline Source/Core/VideoBackends/Vulkan/VulkanContext.cpp]: `u32`

File: Source/Core/DolphinNoGUI/GLFW.h (hunk 2 above)
Comment (linkmauve) [inline Source/Core/DolphinNoGUI/GLFW.h]: This has been copied 1:1 from the X11 version, I will fix them both.

File: Source/Core/VideoBackends/Vulkan/SwapChain.cpp (hunk 4 above)
Comment (linkmauve) [inline Source/Core/VideoBackends/Vulkan/SwapChain.cpp]: No, the `LOG_VULKAN_ERROR()` macro (actually `LogVulkanResult()`) converts res into a string and displays that.

File: Source/Core/DolphinNoGUI/GLFW.h (hunk 7 above)
Comment (linkmauve) [inline Source/Core/DolphinNoGUI/GLFW.h]: The current X11 backend uses a lot of static variables, what would you suggest as the best way to refactor that for multiple backends?

**Comment by Adamillo** [comment]
What's the status on this PR?

**Comment by dreamsyntax** [comment]
This should be able to be closed, per:
https://github.com/dolphin-emu/dolphin/pull/7512#issuecomment-464061562

**Comment by Adamillo** [comment]
This should be closed

**Comment by Rumi-Larry** [comment]
So GLFW has no upsides compared to X11? Or can one built it anyway without dedicated code?

**Comment by dreamsyntax** [comment]
> So GLFW has no upsides compared to X11? Or can one built it anyway without dedicated code?

It was merged (op of this PR commented as such on the other PR)
https://github.com/dolphin-emu/dolphin/pull/7512#issuecomment-464061562

**Comment by phire** [comment]
Closing because it's very outdated. 

On a related note, I have experimental support for a qt platform in my [qtquick branch](https://github.com/phire/dolphin/tree/qtquick), which might (or might not) be the better direction for supporting dolphin-nogui on widows.

**Pull request closed without merging**